simtbx/nanoBragg/nanoBragg_beam.py

```diff
@@ -4,17 +4,52 @@
 from __future__ import print_function, division
 from dxtbx.model.beam import BeamFactory
 from dxtbx_model_ext import flex_Beam
+import numpy as np
+from copy import deepcopy
+
+
+def rotate_axis(v, axis, phi):
+    sinphi = np.sin(phi);
+    cosphi = np.cos(phi);
+    dot = np.dot(axis,v)*(1-cosphi)
+    #dot = (axis[0]*v[0]+axis[1]*v[1]+axis[2]*v[2])*(1.0-cosphi);
+    new_x = axis[0]*dot+v[0]*cosphi+(-axis[2]*v[1]+axis[1]*v[2])*sinphi;
+    new_y = axis[1]*dot+v[1]*cosphi+(+axis[2]*v[0]-axis[0]*v[2])*sinphi;
+    new_z = axis[2]*dot+v[2]*cosphi+(-axis[1]*v[0]+axis[0]*v[1])*sinphi;
+    return new_x, new_y, new_z
 
 
 class NBbeam(object):
 
   def __init__(self):
     self.spectrum = [(1.8, 1e12)] # angstroms, photons per pulse
-    self.unit_s0 = 1, 0, 0
-    self.polarization_fraction = 1
-    self.divergence = 0
-    self.size_mm = 0.001
+    self.unit_s0 = 1, 0, 0  # forward beam direction
+    self.polarization_fraction = 1  # defines horizontal and vertical polarization fraction
+    self.divergence_mrad = 0  # set the divergence cone angle
+    self.divsteps = 0  # number of divergence steps, will be squared (one per horizontal, vertical directions)
+    self.size_mm = 0.001 # beam spot size
     self._undo_nanoBragg_norm_by_nbeams = True # we undo it by default
+    self.prev_xray_beams = None  # used to cache most recent xray_beams property for efficiency
+    self.num_div_angles_within_cone = 0  # used to count how manydivergence angles we sample within the cone of divergence
+
+  @property
+  def divsteps(self):
+    return self._divsteps
+
+  @divsteps.setter
+  def divsteps(self, val):
+    if val > 0:
+      assert val % 2 == 0, "divsteps must be even"
+    self._divsteps = val
+
+  @property
+  def divergences(self):
+    divrange = self.divergence_mrad/1000.
+    if self.divsteps==0:
+      return [(0,0)]
+    else:
+      all_divs = np.arange(0, divrange+1e-7, divrange / self.divsteps) - divrange / 2
+      return [(hdiv, vdiv) for vdiv in all_divs for hdiv in all_divs]
```

Comment:
You might find the expression `np.linspace(-divrange/2, divrange/2, num=self.divsteps)` useful :) ([docs](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html#numpy.linspace)). I also don't know how many divsteps would be usually expected, but `[(hdiv, vdiv) for itertools.product(all_divs, all_divs)]` might be slightly faster and easier to read, though it's probably a matter of preference.

Reply:
As you can clearly see, my priorities are in a strange place, but I decided to comments in case anyone finds these useful. Also, as a pious member of the PEP8 church, there is a bunch of unnecessary semicolons in `rotate_axis`.

Reply:
yea I will clean up rotate_axis, I think there is probably a one-liner in scipy.spatial somewhere

Reply:
(or in scitbx somewhere.. )

Reply:
I accepted your challange and accidentally found one in [C++](https://github.com/cctbx/cctbx_project/blob/ca44a56a6632f52558d9bb049d2ca3e172b5f0a4/simtbx/nanoBragg/nanoBragg.cpp#L4122) and [CUDA](https://github.com/cctbx/cctbx_project/blob/02b6bb6c5bef18f0e54f9a64f51878bf17f07329/simtbx/nanoBragg/nanoBraggCUDA.cu#L957), but not in scitbx. There is `iotbx.detector.rotate_vector_around` which seems to have exactly expected signature that should work with numpy and is exposed to python ([link](https://github.com/cctbx/cctbx_project/blob/bc3b8eb7dfcbe7403cf40110e1ef0dea3a81aab6/iotbx/detectors/detectors_ext.cpp#L317)). This only grounds me in my opinion that there should be a common module with functions like that.

Reply:
yea, I could actually expose the c++ one to python using boost (as opposed to copying it as I did, hence the semicolons).... there is also the route of creating a rotation matrix from axis and angle. I timed it and it took longer though

Reply:
I was looking for a fast way to rotate a point around an axis by an angle. I found some hidden down inside C++ code that was not accessible. I really had to make it faster because it was a bottleneck for me. So I exposed it as mmtbx.reduce.RotatePointDegreesAroundAxisDir() at https://github.com/cctbx/cctbx_project/blob/master/mmtbx/reduce/boost_python/reduce_bpl.cpp#L36 (you can find it used in Python code at https://github.com/cctbx/cctbx_project/blob/master/mmtbx/reduce/Movers.py#L1401). Hope it helps.

Reply:
You can use scitbx.matrix.rotate_point_around_axis if you don't need it to be super fast. The line after the one above that is commented out shows how to call that (which requires adding the offset to the origin).

Reply:
In case anyone is wondering, I looked at scipy and it has a truly elegant interface. `scipy.spatial.transform.Rotation.from_rotvec(np.pi/2 * np.array([0, 0, 1])).as_matrix()` returns a numpy array with rotation matrix. This method can be supplied an array of vectors to provide an array of rotation matrices. I refuse to believe something like this would be slower than python loop.
I felt like this command chain is mostly a tangent, but is there any chance this part of the code would visibly affect performance? If so, I can help with doing numpy magic on 3/4D arrays.

Reply:
Thanks, it seems the mmtbx method offers a significant speedup, I made this change:

```
      vec_xyz = RotatePointDegreesAroundAxisDir((0,0,0), polar_vector, -beam_vector, vdiv*180/np.pi)
      unit_s0 = RotatePointDegreesAroundAxisDir((0,0,0), vert_vector, vec_xyz, hdiv*180/np.pi) 
```

If we converted the outer loop to a large matrix multiply and used the scipy.spatial methods, it might be faster still, but I think for now this is good

Reply:
@bkpoon It may make sense to move RotatePointDegreesAroundAxisDir into scitbx and out of Reduce if it is going to be useful to other projects. I don't mind it being used elsewhere, but I know you like to keep things organized.